lib/fetch/formdata.js

```diff
@@ -0,0 +1,131 @@
+    if (idx === -1) {
+      return null
+    }
+
+    // 2. Return the value of the first entry whose name is name from
+    // this’s entry list.
+    return this[kState][idx].value
+  }
+
+  getAll (name) {
+    // 1. If there is no entry whose name is name in this’s entry list,
+    // then return the empty list.
+    // 2. Return the values of all entries whose name is name, in order,
+    // from this’s entry list.
+    return this[kState]
+      .filter((entry) => entry.name === name)
+      .map((entry) => entry.value)
+  }
+
+  has (name) {
+    // The has(name) method steps are to return true if there is an entry
+    // whose name is name in this’s entry list; otherwise false.
+    return this[kState].findIndex((entry) => entry.name === name) !== -1
+  }
+
+  set (name, value, filename) {
+    // The set(name, value) and set(name, blobValue, filename) method steps
+    // are:
+
+    // 1. Let value be value if given; otherwise blobValue.
+    // 2. Let entry be the result of creating an entry with name, value, and
+    // filename if given.
+    const entry = makeEntry(name, value, filename)
+
+    // 3. If there are entries in this’s entry list whose name is name, then
+    // replace the first such entry with entry and remove the others.
+    const idx = this[kState].findIndex((entry) => entry.name === name)
+    if (idx !== -1) {
+      this[kState][idx] = entry
+      this[kState] = this[kState].filter((entry) => entry.name !== name)
+    } else {
+      // 4. Otherwise, append entry to this’s entry list.
+      this[kState].push(entry)
+    }
+  }
+
+  * [Symbol.iterator] () {
+    // The value pairs to iterate over are this’s entry list’s entries with
+    // the key being the name and the value being the value.
+    for (const { name, value } of this[kState]) {
+      yield [name, value]
+    }
+  }
+}
+
+function makeEntry (name, value, filename) {
+  // To create an entry for name, value, and optionally a filename, run these
+  // steps:
+
+  // Let entry be a new entry.
+  const entry = {
+    name: null,
+    value: null
+  }
+
+  // Set entry’s name to name.
+  entry.name = name
+
+  // If value is a Blob object and not a File object, then set value to a new File
+  // object, representing the same bytes, whose name attribute value is "blob".
+  if (value instanceof Blob && !(value instanceof File)) {
+    value = new File([value], 'blob')
+  }
+
+  // If value is (now) a File object and filename is given, then set value to a
+  // new File object, representing the same bytes, whose name attribute value is
+  // filename.
+  // TODO: This is a bit weird... What if passed value is a File?
+  // Do we just override the name attribute? Since it says "if value is (now)"
+  // does that mean that this lives inside the previous condition? In which case
+  // creating one more File instance doesn't make much sense....
+  if (value instanceof File && filename) {
+    value = new File([value], filename)
```

Comment:
I believe you can merge logic for File and Blob:

```js
if (value instanceof File) {
  // 1. If filename argument is provided then use it as the second argument
  // 2. Else try to use value.name as filename argument
  // 3. If value has no name property (value is Blob), then set filename to "blob"
  value = new File([value], filename || value.name || "blob", {
    type: value.type,
    lastModified: value.lastModified
  })
}
```

Reply:
I'd like to keep the exact wording of the spec.

Reply:
This is fine. It was just a suggestion, because I just thinks it does not make any difference in terms of behaviour.